client/my-sites/checkout/composite-checkout/composite-checkout.tsx

```diff
@@ -263,6 +263,13 @@ export default function CompositeCheckout( {
 		showAddCouponSuccessMessage
 	);
 
+	// Remove the coupon when it doesn't exist on the server-side
+	useEffect( () => {
+		if ( couponStatus === 'applied' && ! couponItem ) {
```

Comment:
This works, but as there's some implicit logic happening in `translateResponseCartToWPCOMCart` that could potentially break this in the future if it is changed, perhaps we should be explicit here:

```suggestion
		if ( couponStatus === 'applied' && responseCart.coupon_savings_total_integer === 0 ) {
```

(It seems to me that the latter condition is sufficient, but if you like you could also check for `! responseCart.coupon`.)